### PR TITLE
Update HTML lang attr on convert

### PR DESCRIPTION
**Pull request by maxmellen**
Hello 新同文堂 maintainers ✨ 

I've been focusing on learning traditional Chinese characters lately, and I've had a great time using the automatic conversion to view sites that don't support 正體中文 in traditional characters.

## Issue 🪲

The only issue I've encountered is that the `html` tag's `lang` attribute doesn't get updated which in a lot of cases causes the browser to dislpay the characters following the [大陸新字型](https://zh.wikipedia.org/wiki/%E6%96%B0%E5%AD%97%E5%BD%A2).

## Proposed solution 🔧

This PR adds a function called in `convertNode` to update the `html` element's `lang` attribute according to the target `LangType`. (only if `lang` is already a valid Chinese language tag)

I'm happy to answer any question or move any code to a better location if needed 🙇‍♀️

## Screenshots 🖼️ 

Here's an example on [cn.vuejs.org](https://cn.vuejs.org):

Please pay attention to the rendering of 進's 走之底 (「⻌」比「⻎」) and 架's lower half (「木」比「朩」), as well as the centered punctuation.

### `lang="zh-Hans"` 简体：
cn vuejs org zh-Hans 简体

### `lang="zh-Hans"` 繁體 (差不多正體) **without this PR**:
cn vuejs org zh-Hans 繁體

### `lang="zh-Hant"` 繁體 (確切正體) **with this PR**:
cn vuejs org zh-Hant 繁體

### CODE REVIEW

**Comment by t7yang** [comment]
hi @maxmellen 

this is a nice improvement 👍 
I didn't notice this before, probably because I forced all Chinese characters to be displayed in Taiwanese font.

I think we should change the `lang` attribute more carefully
some website may rely on the attribute to do something
also, you can force the extension to convert a non Chinese website
and the extension changes the `lang` attribute to `zh-*`
but that's not what the website is supposed to represent.

I think we can introduce a new option to let the user choose whether or not to update the `lang` attribute on convert.
also, we need to ensure changing the `lang` attribute only if
1. the website already has a `lang` attribute and the attribute is one of the valid `zh-*` value.
2. the `lang` attribute is not set, and the browser recognizes the website as a Chinese website.

anyway, I really appreciate it.
feel free to submit more code if you like to implement this feature.
otherwise, I will open an issue to keep track it down and do it when I am free.

**Comment by maxmellen** [comment]
Hello again!

Any update on this?
Happy to update my code if my solution is inadequate.
Would appreciate any feedback.

Thanks :)

**Comment by t7yang** [comment]
sorry, I was busy last week, I will review your code this week, maybe weekend.

File: src/content/convert/convert-nodes.ts [hunk 1]
```diff
@@ -13,6 +14,7 @@ export const convertNode: SConvertNode = async (state, target, nodes) => {
         .then(() => convertNodesText(target, parsedNodes))
         .then(texts => {
           state.mutationObserver?.disconnect();
+          state.updateLang && updateHtmlLang(document.documentElement, target);
```
Comment (t7yang) [inline src/content/convert/convert-nodes.ts:17]: please swap line 17 and 18.
(make sure the lang attribute is only set when the content is successfully updated)

File: src/_locales/zh_TW/messages.json [hunk 2]
```diff
@@ -23,6 +23,9 @@
   "MSG_DYNAMIC_CONVERT": {
     "message": "動態轉換"
   },
+  "MSG_UPDATE_LANG": {
+    "message": "更新頁面語言（讓瀏覽器使用適當的字體）"
```
Comment (t7yang) [inline src/_locales/zh_TW/messages.json:27]: 同時更新網頁的語言屬性（有助於瀏覽器使用適當的字型）
I leave the English description for you :P

**Review by t7yang — CHANGES_REQUESTED** [review]

**Comment by maxmellen** [comment]
Hi @t7yang,

Thanks a lot for your comments! 🙇‍♀️
I’ve updated my branch to address all of your points (see: 44841ddb49191f607bc8a5d2e9018a7e1f67a16a)

Let me know if there’s anything else!
Hope this can be merged soon 🤞

File: src/preference/schema/v2/general.ts [hunk 3]
```diff
@@ -19,5 +19,6 @@ export const generalSchema: Control<PrefGeneral> = dctrl({
   ),
   defaultTarget: vctrl(vldFn({ type: 'string', required: true, enum: [LangType.s2t, LangType.t2s] }), LangType.s2t),
   spaMode: isBoolean(true),
+  updateLang: isBoolean(false),
```
[the remark this answers is not in the document]
Comment (maxmellen) [reply src/preference/schema/v2/general.ts]: Updated in 44841ddb49191f607bc8a5d2e9018a7e1f67a16a

File: src/content/convert/convert-nodes.ts (hunk 1 above)
Comment (maxmellen) [reply src/content/convert/convert-nodes.ts]: Updated in 44841ddb49191f607bc8a5d2e9018a7e1f67a16a

File: src/_locales/zh_TW/messages.json (hunk 2 above)
Comment (maxmellen) [reply src/_locales/zh_TW/messages.json]: Updated in 44841ddb49191f607bc8a5d2e9018a7e1f67a16a

**Review by t7yang — APPROVED** [review]

**Comment by maxmellen** [comment]
Hi @t7yang 👋

I was wondering if there's any chance to see this released any time soon?
Or are you waiting for anything in particular before releasing?

Thanks in advance!

**Comment by t7yang** [comment]
yeah, I am busy on other OSS projects
but I just published the new version on all three browsers
so, stay tuned

**Comment by maxmellen** [comment]
Just got the update on Firefox, great to see this published 💫 
Thanks again 👏

**Comment by t7yang** [comment]
@maxmellen 
btw, I published it even I found a issue, but I know you really want this feature.
I found the Wikipedia doesn't applied the correct font, even the extension changed the `lang` attribute.
if you have time, you can take a look at this issue, PR or further discussion is always welcome :)

**Comment by maxmellen** [comment]
Hey @t7yang 👋 

So I had some time to look into this, and on Wikipedia, the `html` elements's `lang` attribute is indeed updated properly, but as you can see in the screenshot below, basically every element that holds text such as the main content, but also the nav bars and menus have their own `lang` attributes which don't get updated currently.

I've opened a PR that updates the `lang` attribute on every element that already has a Chinese `lang` attribute &rarr; https://github.com/tongwentang/tongwentang-extension/pull/64

Let me know what  you think :)

image

**Pull request merged**